Boxed variable dispatch

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29934 Boxed variable dispatch**

Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

smessmer committed:
Boxed variable dispatch

Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/VariableFallbackKernel.cpp:
```diff
@@ -0,0 +1,40 @@
+
+void variable_fallback_kernel(const OperatorHandle& op, Stack* stack) {
+    at::AutoNonVariableTypeMode _var_guard(true);
+    Dispatcher::singleton().callBoxed(op, stack);
```
OK, this isn't really what I've been advocating for, but I'll let it slide for now.

smessmer replied:
it's not? IIRC, we said we're setting the thread local flag and re-dispatching. That's what I'm doing here. What would you do instead?

ezyang replied:
We also need to setup `grad_fn` on the outputs to error if you try to backprop through them! Otherwise you'll just poop out `requires_grad=False` outputs which will silently fail to backprop if someone uses them.

smessmer replied:
Ah, that's what you mean. Yes, we should do that. This doesn't regress behavior though, it's just as broken as it was before. There were some issues with adding the grad_fn if I remember correctly which meant we can't do it right now, but I'll look at it in a separate PR.

ezyang replied:
Yep, that's why I approved.

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]

smessmer committed:
Update on "Boxed variable dispatch"


Previously, when doing boxed dispatch (e.g. custom ops), the dispatcher manually removed the VariableTensorId flag before dispatching
because custom ops don't have variable kernels.
This is one of the blockers that prevented us from using the boxed dispatch mechanism for ops from native_functions.yaml because they define variable kernels and need them to be called for autograd.

This PR changes that. The dispatcher doesn't remove the VariableTensorId flag anymore.
Instead, to make custom ops work, we implement a variable fallback kernel that is called whenever no other variable kernel was found.

See https://fb.quip.com/czCUA1NzMeEt for the detailed plan.

Differential Revision: [D18542342](https://our.internmc.facebook.com/intern/diff/D18542342/)

[ghstack-poisoned]
